gengo/src/lib.rs

```diff
@@ -39,127 +42,243 @@ const GLOB_MATCH_OPTIONS: MatchOptions = MatchOptions {
 
 /// The main entry point for Gengo.
 pub struct Gengo {
-    repository: Repository,
+    repository: gix::ThreadSafeRepository,
     analyzers: Analyzers,
     read_limit: usize,
     documentation: Documentation,
     generated: Generated,
     vendored: Vendored,
 }
 
-// TODO parse .gitattributes to get language overrides.
-impl Gengo {
-    const ATTR_CHECK_FLAGS: [AttrCheckFlags; 2] =
-        [AttrCheckFlags::NO_SYSTEM, AttrCheckFlags::INDEX_THEN_FILE];
-    /// Resolves a revision to a commit.
-    fn rev(&self, rev: &str) -> Result<Commit> {
-        let reference = self.repository.revparse_single(rev)?;
-        let commit = reference.peel_to_commit()?;
-        Ok(commit)
+#[derive(Clone)]
+struct GitState {
+    attr_stack: gix::worktree::Stack,
+    attr_matches: gix::attrs::search::Outcome,
+}
+
+impl GitState {
+    fn new(repo: &gix::Repository, tree_id: &gix::oid) -> Result<(Self, gix::index::State)> {
+        let index = repo.index_from_tree(tree_id)?;
+        let attr_stack = repo.attributes_only(
+            &index,
+            gix::worktree::stack::state::attributes::Source::IdMapping,
+        )?;
+        let attr_matches = attr_stack.selected_attribute_matches([
+            "gengo-language",
+            "gengo-generated",
+            "gengo-documentation",
+            "gengo-vendored",
+            "gengo-detectable",
+        ]);
+        Ok((
+            Self {
+                attr_stack,
+                attr_matches,
+            },
+            index.into_parts().0,
+        ))
+    }
+}
+
+struct BlobEntry {
+    // Just for path and id access
+    index_entry: gix::index::Entry,
+    result: Option<Entry>,
+}
+
+/// The result of analyzing a repository or a single submodule
+struct Results {
+    /// If this is a submodule, the root is not empty and the full path to where our paths start.
+    root: BString,
+    entries: Vec<BlobEntry>,
+    path_storage: gix::index::PathStorage,
+}
+
+impl Results {
+    /// Create a data structure that holds index entries as well as our results per entry.
+    /// Return a list of paths at which submodules can be found, along with their
+    /// commit ids.
+    fn from_index(
+        root: BString,
+        index: gix::index::State,
+    ) -> (Self, Vec<(BString, gix::ObjectId)>) {
+        let (entries, path_storage) = index.into_entries();
+        let mut submodules = Vec::new();
+        let entries: Vec<_> = entries
+            .into_iter()
+            .filter(|e| {
+                if e.mode == gix::index::entry::Mode::COMMIT {
```

Comment:
Nitpick: Import `gix::index::entry::Mode` so this doesn't need to be so verbose.
Maybe make this a `match` since it's `e.mode == X` repeated.